delivery_purchase_label/models/purchase_order.py

```diff
@@ -0,0 +1,132 @@
+        "Delivery Label Picking",
+        store=True,
+        readonly=True,
+    )
+
+    @api.model
+    def _states_to_generate_delivery_label(self):
+        """Labels will be (re)generated only if the PO is in one of these states."""
+        return ["draft", "sent"]
+
+    @api.model
+    def _mail_templates_to_not_attach_labels(self):
+        return self.env.ref("purchase.email_template_edi_purchase_reminder").ids
+
+    def action_rfq_send(self):
+        self.ensure_one()
+        self._generate_purchase_delivery_label()
+        return super().action_rfq_send()
+
+    def _is_valid_for_vendor_labels(self):
+        self.ensure_one()
+        if self.state not in self._states_to_generate_delivery_label():
+            return
+        if not self.dest_address_id:
+            return False
+        if not any(
+            product.type in ["product", "consu"]
+            for product in self.order_line.product_id
+        ):
+            return False
+        return True
+
+    def _is_picking_label_uptodate(self):
+        """Check if the picking label needs to be regenerated."""
+        self.ensure_one()
+        if not self.delivery_label_picking_id:
+            return False
+        pick = self.delivery_label_picking_id
+        moves_new = []
+        for line in self.order_line:
+            for value in line._prepare_stock_moves(self.delivery_label_picking_id):
+                moves_new.append(
+                    {key: value[key] for key in ("product_id", "product_uom_qty")}
+                )
+        move_prev = [
+            {"product_id": move.product_id.id, "product_uom_qty": move.product_uom_qty}
+            for move in pick.move_lines
+        ]
+        if len(move_prev) == len(moves_new):
+            if [
+                move_value for move_value in move_prev if move_value not in moves_new
+            ] == []:
+                return True
+        return False
+
+    def _generate_purchase_delivery_label(self):
+        """Create a transfer to generate the carrier labels."""
+        self.ensure_one()
+        if not self._is_valid_for_vendor_labels():
+            return
+        # Find the carrier that will be used
+        carrier = self.partner_id.purchase_delivery_carrier_id
+        if not carrier.purchase_label_picking_type:
+            return
+        if not self.partner_id.property_stock_supplier.id:
+            raise UserError(
+                _(
+                    "You must set a Vendor Location for this partner %s",
+                    self.partner_id.name,
+                )
+            )
+        if self._is_picking_label_uptodate():
+            return
+        order = self.with_company(self.company_id)
+        # Create and process the transer to send the labels
+        values = order._get_purchase_delivery_label_picking_value(carrier)
+        picking = self.env["stock.picking"].with_user(SUPERUSER_ID).create(values)
+        moves = order.order_line._create_stock_moves(picking)
```

Comment:
This is exactly what is causing problems now. Without the dry-run you are comparing the ordered products and qties against a maybe different transferred products. The moves are not the same as the ordered for example for products with a kit bom.

Reply:
Knowing that the `mrp` module will change the moves when the procurement is being run. This is probably too complex to reproduce here.
Would checking  that the purchase lines have not changed (using the `write_date`) since the current delivery label picking has been generated be a good enough solution ?

Reply:
Finally using a savepoint